distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/SyncPluginsCommand.java

```diff
@@ -0,0 +1,188 @@
+
+        execute(terminal, env, isBatch, isPurge, isDry);
+    }
+
+    protected void execute(Terminal terminal, Environment env, boolean isBatch, boolean isPurge, boolean isDry) throws Exception {
+        final RemovePluginAction removePluginAction = new RemovePluginAction(terminal, env, isPurge);
+        final InstallPluginAction installPluginAction = new InstallPluginAction(terminal, env, isBatch);
+
+        execute(terminal, env, isDry, removePluginAction, installPluginAction);
+    }
+
+    protected void execute(
+        Terminal terminal,
+        Environment env,
+        boolean isDry,
+        RemovePluginAction removePluginAction,
+        InstallPluginAction installPluginAction
+    ) throws Exception {
+        if (Files.exists(env.pluginsFile()) == false) {
+            throw new UserException(ExitCodes.CONFIG, "Plugins directory missing: " + env.pluginsFile());
+        }
+
+        // 1. Parse descriptor file
+        final PluginsManifest pluginsManifest = PluginsManifest.parseManifest(env);
+
+        // 2. Get list of installed plugins
+        final List<PluginInfo> existingPlugins = getExistingPlugins(env, terminal);
+
+        // 3. Calculate changes
+        final List<PluginDescriptor> pluginsThatShouldExist = pluginsManifest.getPlugins();
+        final List<PluginDescriptor> pluginsThatActuallyExist = existingPlugins.stream()
+            .map(info -> new PluginDescriptor(info.getName()))
+            .collect(Collectors.toList());
+
+        final List<PluginDescriptor> pluginsToInstall = difference(pluginsThatShouldExist, pluginsThatActuallyExist);
+        final List<PluginDescriptor> pluginsToRemove = difference(pluginsThatActuallyExist, pluginsThatShouldExist);
+
+        printRequiredChanges(terminal, isDry, pluginsToRemove, pluginsToInstall);
+
+        if (isDry) {
+            return;
+        }
+
+        // 5. Remove any plugins that are not in the descriptor
+        if (pluginsToRemove.isEmpty() == false) {
+            removePluginAction.execute(pluginsToRemove);
+        }
+
+        // 6. Add any plugins that are in the descriptor but missing from disk
+        if (pluginsToInstall.isEmpty() == false) {
+            installPluginAction.setProxy(buildProxy(pluginsManifest.getProxy()));
+            installPluginAction.execute(pluginsToInstall);
+        }
+    }
+
+    private List<PluginInfo> getExistingPlugins(Environment env, Terminal terminal) throws IOException {
+        final List<PluginInfo> plugins = new ArrayList<>();
+
+        try (DirectoryStream<Path> paths = Files.newDirectoryStream(env.pluginsFile())) {
+            for (Path pluginPath : paths) {
+                PluginInfo info = PluginInfo.readFromProperties(env.pluginsFile().resolve(pluginPath));
+                plugins.add(info);
+                if (info.getElasticsearchVersion().equals(Version.CURRENT) == false) {
+                    terminal.errorPrintln(
+                        "WARNING: plugin ["
```

Comment:
Should we really be doing this here? I'm sure we validate this elsewhere, no?

Reply:
We also do this check in the `list` subcommand, but you might conceivably never run that if you were relying on the `sync` command. Also, I think when I first added it I thought that we might eventually remove the imperative commands altogether. The `list` command is the only other place (besides Elasticsearch itself) that we actually inspect the currently-installed plugins.